File: GameEngines/Cocos2d-x/cocos2d/extensions/DegenHF/AuthExtension.cpp
#include "AuthExtension.h"
#include <cocos2d.h>

USING_NS_CC;

namespace DegenHF {
namespace Cocos2dx {

static AuthExtension* s_instance = nullptr;

AuthExtension* AuthExtension::getInstance() {
    if (!s_instance) {
        s_instance = new AuthExtension();
    }
    return s_instance;
}

AuthExtension::AuthExtension()
    : m_authHandler(nullptr)
    , m_initialized(false) {
}

AuthExtension::~AuthExtension() {
    if (m_authHandler) {
        m_authHandler->saveAuthData();
    }
}

bool AuthExtension::init(const ECC::AuthHandler::Config& config) {
    if (m_initialized) {
        return true;
    }

    m_authHandler = std::make_unique<ECC::AuthHandler>(config);
    if (!m_authHandler->initialize()) {
        CCLOG("Failed to initialize ECC Auth Handler");
        return false;
    }

    m_initialized = true;
    CCLOG("DegenHF Cocos2d-x Auth Extension initialized successfully");
    return true;
}

void AuthExtension::registerUser(const std::string& username, const std::string& password, RegisterCallback callback) {
    if (!m_initialized || !m_authHandler) {
        if (callback) {
            runCallbackOnMainThread([callback]() {
                callback(false, "", "Extension not initialized");
            });
        }
        return;
    }

    // Perform registration on background thread
    Director::getInstance()->getScheduler()->performFunctionInCocosThread([this, username, password, callback]() {
        auto result = m_authHandler->registerUser(username, password);

        if (callback) {
            runCallbackOnMainThread([callback, result]() {
                callback(result.success, result.userId, result.errorMessage);
            });
        }
    });
}

void AuthExtension::loginUser(const std::string& username, const std::string& password, LoginCallback callback) {
    if (!m_initialized || !m_authHandler) {
        if (callback) {
            runCallbackOnMainThread([callback]() {
                callback(false, "", "", "", "Extension not initialized");
            });
        }
        return;
    }

    // Perform login on background thread
    Director::getInstance()->getScheduler()->performFunctionInCocosThread([this, username, password, callback]() {
        auto result = m_authHandler->authenticateUser(username, password);

        if (callback) {
            runCallbackOnMainThread([callback, result]() {
                callback(result.success, result.token, result.userId, result.username, result.errorMessage);
            });
        }
    });
}

void AuthExtension::logoutUser(AuthCallback callback) {
    if (!m_initialized || !m_authHandler) {
        if (callback) {
            runCallbackOnMainThread([callback]() {
                callback(false, "Extension not initialized");
            });
        }
        return;
    }

    m_authHandler->logout();
    m_authHandler->saveAuthData();

    if (callback) {
        runCallbackOnMainThread([callback]() {
            callback(true, "Logged out successfully");
        });
    }
}

bool AuthExtension::isLoggedIn() const {
    if (!m_initialized || !m_authHandler) {
        return false;
    }
    return m_authHandler->isUserLoggedIn();
}

std::string AuthExtension::getCurrentUserId() const {
    if (!m_initialized || !m_authHandler) {
        return "";
    }
    return m_authHandler->getCurrentUserId();
}

std::string AuthExtension::getCurrentUsername() const {
    if (!m_initialized || !m_authHandler) {
        return "";
    }
    return m_authHandler->getCurrentUsername();
}

void AuthExtension::verifyToken(const std::string& token, AuthCallback callback) {
    if (!m_initialized || !m_authHandler) {
        if (callback) {
            runCallbackOnMainThread([callback]() {
                callback(false, "Extension not initialized");
            });
        }
        return;
    }

    // Perform verification on background thread
    Director::getInstance()->getScheduler()->performFunctionInCocosThread([this, token, callback]() {
        auto result = m_authHandler->verifyToken(token);

        if (callback) {
            runCallbackOnMainThread([callback, result]() {
                callback(result.valid, result.valid ? "Token valid" : result.errorMessage);
            });
        }
    });
}

void AuthExtension::createSession(std::function<void(const std::string& sessionId)> callback) {
    if (!m_initialized || !m_authHandler) {
        if (callback) {
            runCallbackOnMainThread([callback]() {
                callback("");
            });
        }
        return;
    }

    std::string userId = m_authHandler->getCurrentUserId();
    if (userId.empty()) {
        if (callback) {
            runCallbackOnMainThread([callback]() {
                callback("");
            });
        }
        return;
    }

    std::string sessionId = m_authHandler->createSession(userId);

    if (callback) {
        runCallbackOnMainThread([callback, sessionId]() {
            callback(sessionId);
        });
    }
}

void AuthExtension::getSessionInfo(const std::string& sessionId,
                                  std::function<void(bool valid, const std::string& userId, const std::string& username)> callback) {
    if (!m_initialized || !m_authHandler) {
        if (callback) {
            runCallbackOnMainThread([callback]() {
                callback(false, "", "");
            });
        }
        return;
    }

    std::string userId, username;
    bool valid = m_authHandler->getSession(sessionId, userId, username);

    if (callback) {
        runCallbackOnMainThread([callback, valid, userId, username]() {
            callback(valid, userId, username);
        });
    }
}

void AuthExtension::saveAuthState() {
    if (m_initialized && m_authHandler) {
        m_authHandler->saveAuthData();
    }
}

void AuthExtension::loadAuthState() {
    if (m_initialized && m_authHandler) {
        m_authHandler->loadAuthData();
    }
}

void AuthExtension::runCallbackOnMainThread(std::function<void()> callback) {
    if (callback) {
        Director::getInstance()->getScheduler()->performFunctionInCocosThread(callback);
    }
}

} // namespace Cocos2dx
} // namespace DegenHF

File: GameEngines/Godot/addons/DegenHF/AuthExtension.cs
#if GODOT
using Godot;
using System;
using System.Collections.Generic;
using System.Threading.Tasks;

namespace DegenHF.Godot
{
    /// <summary>
    /// Godot extension for DegenHF ECC authentication
    ///
    /// Provides easy-to-use authentication integration for Godot games
    /// with automatic scene management and GDScript callbacks.
    /// </summary>
    [Tool]
    public class AuthExtension : Node
    {
        /// <summary>
        /// Authentication callback types
        /// </summary>
        public delegate void AuthCallback(bool success, string message);
        public delegate void RegisterCallback(bool success, string userId, string message);
        public delegate void LoginCallback(bool success, string token, string userId, string username, string message);

        private static AuthExtension _instance;
        private ECC.AuthHandler _authHandler;
        private bool _initialized;

        /// <summary>
        /// Get singleton instance
        /// </summary>
        /// <returns>AuthExtension instance</returns>
        public static AuthExtension GetInstance()
        {
            if (_instance == null)
            {
                _instance = new AuthExtension();
            }
            return _instance;
        }

        public override void _Ready()
        {
            if (_instance == null)
            {
                _instance = this;
            }
        }

        /// <summary>
        /// Initialize the extension
        /// </summary>
        /// <param name="config">Authentication configuration</param>
        /// <returns>true if initialization successful</returns>
        public bool Init(ECC.AuthHandler.Config config = null)
        {
            if (_initialized)
            {
                return true;
            }

            _authHandler = new ECC.AuthHandler(config);
            if (!_authHandler.Initialize())
            {
                GD.PrintErr("Failed to initialize ECC Auth Handler");
                return false;
            }

            _initialized = true;
            GD.Print("DegenHF Godot Auth Extension initialized successfully");
            return true;
        }

        /// <summary>
        /// Register a new user
        /// </summary>
        /// <param name="username">Username for registration</param>
        /// <param name="password">Password for registration</param>
        /// <param name="callback">Callback function for result</param>
        public void RegisterUser(string username, string password, RegisterCallback callback = null)
        {
            if (!_initialized || _authHandler == null)
            {
                callback?.Invoke(false, "", "Extension not initialized");
                return;
            }

            // Perform registration asynchronously
            Task.Run(() =>
            {
                var result = _authHandler.RegisterUser(username, password);
                CallDeferred("emit_signal", "register_completed", result.Success, result.UserId, result.ErrorMessage);

                if (callback != null)
                {
                    CallDeferred("invoke_callback", callback.Method.Name, result.Success, result.UserId, result.ErrorMessage);
                }
            });
        }

        /// <summary>
        /// Login user
        /// </summary>
        /// <param name="username">Username for login</param>
        /// <param name="password">Password for login</param>
        /// <param name="callback">Callback function for result</param>
        public void LoginUser(string username, string password, LoginCallback callback = null)
        {
            if (!_initialized || _authHandler == null)
            {
                callback?.Invoke(false, "", "", "", "Extension not initialized");
                return;
            }

            // Perform login asynchronously
            Task.Run(() =>
            {
                var result = _authHandler.AuthenticateUser(username, password);
                CallDeferred("emit_signal", "login_completed", result.Success, result.Token, result.UserId, result.Username, result.ErrorMessage);

                if (callback != null)
                {
                    CallDeferred("invoke_callback", callback.Method.Name, result.Success, result.Token, result.UserId, result.Username, result.ErrorMessage);
                }
            });
        }

        /// <summary>
        /// Logout current user
        /// </summary>
        /// <param name="callback">Callback function for result</param>
        public void LogoutUser(AuthCallback callback = null)
        {
            if (!_initialized || _authHandler == null)
            {
                callback?.Invoke(false, "Extension not initialized");
                return;
            }

            _authHandler.Logout();
            _authHandler.SaveAuthData();

            callback?.Invoke(true, "Logged out successfully");
        }

        /// <summary>
        /// Check if user is logged in
        /// </summary>
        /// <returns>true if user is logged in</returns>
        public bool IsLoggedIn()
        {
            if (!_initialized || _authHandler == null)
            {
                return false;
            }
            return _authHandler.IsUserLoggedIn();
        }

        /// <summary>
        /// Get current user ID
        /// </summary>
        /// <returns>Current user ID or empty string</returns>
        public string GetCurrentUserId()
        {
            if (!_initialized || _authHandler == null)
            {
                return "";
            }
            return _authHandler.GetCurrentUserId();
        }

        /// <summary>
        /// Get current username
        /// </summary>
        /// <returns>Current username or empty string</returns>
        public string GetCurrentUsername()
        {
            if (!_initialized || _authHandler == null)
            {
                return "";
            }
            return _authHandler.GetCurrentUsername();
        }

        /// <summary>
        /// Verify authentication token
        /// </summary>
        /// <param name="token">Token to verify</param>
        /// <param name="callback">Callback function for result</param>
        public void VerifyToken(string token, AuthCallback callback = null)
        {
            if (!_initialized || _authHandler == null)
            {
                callback?.Invoke(false, "Extension not initialized");
                return;
            }

            // Perform verification asynchronously
            Task.Run(() =>
            {
                var result = _authHandler.VerifyToken(token);
                CallDeferred("emit_signal", "verify_completed", result.Valid, result.Valid ? "Token valid" : result.ErrorMessage);

                if (callback != null)
                {
                    CallDeferred("invoke_callback", callback.Method.Name, result.Valid, result.Valid ? "Token valid" : result.ErrorMessage);
                }
            });
        }

        /// <summary>
        /// Create a secure session
        /// </summary>
        /// <param name="callback">Callback function with session ID</param>
        public void CreateSession(Action<string> callback = null)
        {
            if (!_initialized || _authHandler == null)
            {
                callback?.Invoke("");
                return;
            }

            string userId = _authHandler.GetCurrentUserId();
            if (string.IsNullOrEmpty(userId))
            {
                callback?.Invoke("");
                return;
            }

            string sessionId = _authHandler.CreateSession(userId);
            callback?.Invoke(sessionId);
        }

        /// <summary>
        /// Get session information
        /// </summary>
        /// <param name="sessionId">Session ID to query</param>
        /// <param name="callback">Callback function with user info</param>
        public void GetSessionInfo(string sessionId, Action<bool, string, string> callback = null)
        {
            if (!_initialized || _authHandler == null)
            {
                callback?.Invoke(false, "", "");
                return;
            }

            string userId, username;
            bool valid = _authHandler.GetSession(sessionId, out userId, out username);
            callback?.Invoke(valid, userId, username);
        }

        /// <summary>
        /// Save authentication state
        /// </summary>
        public void SaveAuthState()
        {
            if (_initialized && _authHandler != null)
            {
                _authHandler.SaveAuthData();
            }
        }

        /// <summary>
        /// Load authentication state
        /// </summary>
        public void LoadAuthState()
        {
            if (_initialized && _authHandler != null)
            {
                _authHandler.LoadAuthData();
            }
        }

        /// <summary>
        /// Get the underlying auth handler
        /// </summary>
        /// <returns>Pointer to ECC auth handler</returns>
        public ECC.AuthHandler GetAuthHandler()
        {
            return _authHandler;
        }

        // Signal definitions for GDScript
        [Signal]
        public delegate void RegisterCompletedEventHandler(bool success, string userId, string message);

        [Signal]
        public delegate void LoginCompletedEventHandler(bool success, string token, string userId, string username, string message);

        [Signal]
        public delegate void VerifyCompletedEventHandler(bool valid, string message);

        // Helper method for deferred callback invocation
        private void InvokeCallback(string methodName, params object[] args)
        {
            // This method is called via CallDeferred to ensure thread safety
            // The actual callback invocation would be handled by the specific method
        }
    }
}
#endif